documentation for systemUser methods in Request Context (#494)
cap-js/docs: java/request-contexts.md

[JwtTokenAuthenticationInfo](https://www.javadoc.io/doc/com.sap.cds/cds-services-api/latest/com/sap/cds/services/authentication/JwtTokenAuthenticationInfo.html) represents a JWT token, but [BasicAuthenticationInfo](https://www.javadoc.io/doc/com.sap.cds/cds-services-api/latest/com/sap/cds/services/authentication/BasicAuthenticationInfo.html) can be observed on requests with basic authentication (e.g. test scenario with mock users). The method `as(Class<? extends AuthenticationInfo>)` helps to perform the downcast to a concrete subtype.

## Defining New Request Contexts { #defining-requestcontext}

The CAP Java SDK allows you to create new Request Contexts and define their scope. This helps you to control, which set of parameters is used when events are processed by services.

There are a few typical use cases in a CAP-based, multitenant application on SAP BTP in which creation of new Request Contexts is necessary. These scenarios are identified by a combination of the user (technical or named) and the tenant (provider or subscribed).

> <img src="./assets/requestcontext.drawio.svg"  alt="A named user can switch to a technical user in the same/subscriber tenant using the systemUser() method. Also, a named user can switch to a technical user in the provider tenant using the systemUserProvider() method. In addition technical users provider/subscriber tenants can switch to technical users on provider/subscriber tenants using the methods systemUserProvider() or systemUser(tenant). ">

Reviewer: Hi @StefanHenke ,

pls have a look especially at this one. I tried to map it to the scenarios that follow. Maybe it limits it's meaning but it's more tangible for me. Let me know if you'd like to revert it to the old one or if this is fine.

Author: @renejeglinsky : The good thing with your diagram is that it names the different scenarios. However, the third row `switching to a specific tenant` includes the call `systemUserProvider()`. This actually would need to be moved to the second row `switching to provider tenant` which then requires a third box in this row. Not sure if it is then still fine.

I gathered some feedback from a colleague in the team. He got the message from both diagrams.

Reviewer: How about that?
![image](https://github.com/cap-js/docs/assets/32062331/8a924fcf-f4ca-4a08-ad2d-8b521370a65a)

Author: Not 100% convinced :) But let´s go with it. If we receive feedback, we can always enhance.

| anonymousUser()      | Switches to an anonymous user.                                                                                                       |
| privilegedUser()     | Elevates the current `UserInfo` to by-pass all authorization checks.                                                                 |

::: info Note
The [RequestContextRunner](https://www.javadoc.io/doc/com.sap.cds/cds-services-api/latest/com/sap/cds/services/runtime/RequestContextRunner.html) API does not allow you to create a Request Context based on a named user. Named user contexts are only created by the CAP Java framework as initial Request Context is based on appropriate authentication information (e.g. JWT token) attached to the incoming HTTP request.
:::

In the following a few concrete examples are given:
- [Switching to Technical User](#switching-to-technical-user)
- [Switching to Provider Tenant](#switching-to-provider-tenant)
- [Switching to a Specific Technical Tenant](#switching-to-a-specific-technical-tenant)

### Switching to Technical User

<img src="./assets/nameduser.drawio.svg"  alt="The graphic is explained in the accompanying text.">

> The incoming JWT token triggers the creation of an initial RequestContext with a named user. Accesses to the database in the OData Adapter as well as the custom `on` handler are executed within <i>tenant1</i> and authorization checks are performed for user <i>JohnDoe</i>. An additionally defined `after` handler wants to call out to an external service without propagating the named user <i>JohnDoe</i>.

Reviewer: Is it John Doe or is this the technical name?
```suggestion
The incoming JWT token triggers the creation of an initial RequestContext with a named user. Accesses to the database in the OData Adapter as well as the custom `on` handler are executed within <i>tenant1</i> and authorization checks are performed for user <i>John Doe</i>. An additionally defined `after` handler wants to call out to an external service without propagating the named user <i>John Doe</i>.
```

Author: It is correct with John Doe. Only the after handler will be executed with a technical user. I made the last sentence a bit clearer.

[JwtTokenAuthenticationInfo](https://www.javadoc.io/doc/com.sap.cds/cds-services-api/latest/com/sap/cds/services/authentication/JwtTokenAuthenticationInfo.html) represents a JWT token, whereas [BasicAuthenticationInfo](https://www.javadoc.io/doc/com.sap.cds/cds-services-api/latest/com/sap/cds/services/authentication/BasicAuthenticationInfo.html) can be observed on requests with basic authentication (e.g. test scenario with mock users). The method `as(Class<? extends AuthenticationInfo>)` helps to perform the downcast to a concrete subtype.

## Defining new Request Contexts { #defining-requestcontext}

The CAP Java SDK allows you to create new Request Contexts and define their scope. This helps you to control, which set of parameters is used when events are processed by services.

There are a few typical use cases in a CAP-based, multi-tenant application on BTP in which creation of new Request Contexts is necessary. These scenarios are identified by a combination of the user (technical or named) and the tenant (provider or subscribed).

> <img src="./assets/requestcontext.png" width="300px">

Reviewer: You can use the svg file here. There's no need to create and use a png. Nice, right? :)

Author: That´s great :) I didn´t know that. Let me change it

In the following a few concrete examples are given:

1) Switch to technical user
<img src="./assets/nameduser.drawio.svg" width="500px">

The incoming JWT token will trigger creation of an initial RequestContext with a named user. Accesses to the database in the OData Adapter as well as the custom `on` handler will be executed within <i>tenant1</i> and authorization checks are performed for user <i>JohnDoe</i>. An additionally defined `after` handler wants to call out to an external service without propagating the named user <i>JohnDoe</i>. 
Therefore, the `after` handler needs to create a new Request Context. To achieve this, it is required to call `requestContext()` on the current `CdsRuntime` and use the `systemUser()` method to remove the named user from the new Request Context. The code being executed in the passed `java.util.function.Function` is triggered with the `run()` method.   

```java
@After(entity = Books_.CDS_NAME)
public void afterHandler(EventContext context){

>     runtime.requestContext().systemUser().run(reqContext -> {

Reviewer 2: ```suggestion
    runtime.requestContext().systemUserProvider().run(reqContext -> {
```

Reviewer 2: Nice docu. I'd vote for showing the systemUserProvider() flow as important scenario.

        // call technical service
        ...
    });
}
```

2) Switch to a specific tenant
<img src="./assets/switchtenant.drawio.svg" width="400px">

The application is using a job scheduler which needs to regularly perform tasks on behalf of a certain tenant. By default, background executions (e.g. in a dedicated thread pool) are not associated to any subscriber tenant and user. In this case, it is necessary to explicitly define a new Request Context based on the subscribed tenant by calling `systemUser(tenantId)`. This will ensure that the Persistence Service will perform the query for the specified tenant.      

```java

> runtime.requestContext().systemUserSubscriber(tenant).run(reqContext -> {

Reviewer 2: ```suggestion
runtime.requestContext().systemUser(tenant).run(reqContext -> {
```
